File: 02-longest-subarray-sum-k/solution.java
// Longest Subarray with Sum K
// Time: O(n), Space: O(n)

import java.util.HashMap;
import java.util.Map;

public class Solution {
    public int longestSubarrayWithSumK(int[] arr, int k) {
        Map<Integer, Integer> prefixSumMap = new HashMap<>();
        int sum = 0;
        int maxLen = 0;
        
        for (int i = 0; i < arr.length; i++) {
            sum += arr[i];
            
            if (sum == k) {
                maxLen = i + 1;
            }
            
            if (prefixSumMap.containsKey(sum - k)) {
                maxLen = Math.max(maxLen, i - prefixSumMap.get(sum - k));
            }
            
            if (!prefixSumMap.containsKey(sum)) {
                prefixSumMap.put(sum, i);
            }
        }
        
        return maxLen;
    }
    
    // Test
    public static void main(String[] args) {
        Solution sol = new Solution();
        System.out.println(sol.longestSubarrayWithSumK(new int[]{1, 2, 3, 1, 1, 1, 1}, 3)); // Output: 3
    }
}

File: 02-longest-subarray-sum-k/solution.cs
// Longest Subarray with Sum K
// Time: O(n), Space: O(n)

using System;
using System.Collections.Generic;

public class Solution {
    public int LongestSubarrayWithSumK(int[] arr, int k) {
        Dictionary<int, int> prefixSumMap = new Dictionary<int, int>();
        int sum = 0;
        int maxLen = 0;
        
        for (int i = 0; i < arr.Length; i++) {
            sum += arr[i];
            
            if (sum == k) {
                maxLen = i + 1;
            }
            
            if (prefixSumMap.ContainsKey(sum - k)) {
                maxLen = Math.Max(maxLen, i - prefixSumMap[sum - k]);
            }
            
            if (!prefixSumMap.ContainsKey(sum)) {
                prefixSumMap[sum] = i;
            }
        }
        
        return maxLen;
    }
    
    // Test
    public static void Main(string[] args) {
        Solution sol = new Solution();
        Console.WriteLine(sol.LongestSubarrayWithSumK(new int[] {1, 2, 3, 1, 1, 1, 1}, 3)); // Output: 3
    }
}
